acbuild/script.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2016 The appc Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var (
+	cmdScript = &cobra.Command{
+		Use:     "script SCRIPT_FILE",
+		Short:   "Runs an acbuild script",
+		Example: "acbuild script build-myapp.acb",
+		Run:     runWrapper(runScript),
+	}
+)
+
+func init() {
+	cmdAcbuild.AddCommand(cmdScript)
+}
+
+func runScript(cmd *cobra.Command, args []string) (exit int) {
+	if len(args) != 1 {
+		cmd.Usage()
+		return 1
+	}
+
+	scriptName := args[0]
+	rawScript, err := ioutil.ReadFile(args[0])
+	if err != nil {
+		stderr("script: %v", err)
+		return getErrorCode(err)
+	}
+
+	if debug {
+		stderr("Running script from %s", scriptName)
+	}
+
+	err = execScript(rawScript)
+	if err != nil {
+		stderr("script: %v", err)
+		return getErrorCode(err)
+	}
+	return 0
+}
+
+func execScript(rawScript []byte) error {
+	script := strings.Split(string(rawScript), "\n")
+	for i := 0; i < len(script); i++ {
+		script[i] = trimComments(script[i])
+	}
+
+	for _, line := range script {
+		if strings.HasPrefix(line, "run") && os.Geteuid() != 0 {
+			return fmt.Errorf("scripts using the run subcommand must be run as root")
+		}
+	}
+
+	tmpDir, err := ioutil.TempDir("", "acbuild")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+	contextpath = tmpDir
+
+	for _, line := range script {
+		if line != "" {
+			err := execACBuild(tmpDir, line)
+			if err != nil {
+				if !strings.HasPrefix(line, "begin") {
+					err1 := newACBuild().End()
+					if err1 != nil {
+						stderr("script: %v", err1)
+					}
+				}
+				return err
+			}
+		}
+	}
+	err = newACBuild().End()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func execACBuild(workPath, line string) error {
+	args := []string{"--debug", "--work-path=" + workPath}
+	args = append(args, tokenizeLine(line)...)
```

Comment:
I think it may be worth having some further validation here and check that argv[1] is a valid subcommand. I think you may also want to check for the "script in script" (is it a valid usecase?).